### PR TITLE
Refactor `CalibrationModel` towards concurrent univariate/multivariate compatibility

### PR DESCRIPTION
**Pull request by michaelosthege**
+ `CalibrationModel.ndim` attribute was added
+ `infer_independent` logic was extracted into a local function, paving the way for future multivariate `infer_independent` implementations.
+ Some exception tests were made more accurate (matching messages) and a typing bug that was found through that was fixed.
+ `NumericPosterior` was deprecated/renamed to `UnivariateInferenceResult`, inheriting from a superclass `InferenceResult` whoose signature is compatible with arbitrary dimensionality.

### CODE REVIEW

File: calibr8/core.py [hunk 1]
```diff
@@ -51,52 +51,111 @@ def __init__(
...
-    def eti_lower(self) -> float:
+    def eti_pdf(self) -> numpy.ndarray:
+        """Values of the posterior probability density at the positions `eti_x`."""
+        return self._eti_pdf
+
+    @property
+    def eti_lower(self) -> Union[float, numpy.ndarray]:
         """ Lower bound of the ETI. This is the first value in `eti_x`. """
-        return self.eti_x[0]
+        return self._eti_x[..., 0]
```
Comment (lhelleckes) [inline calibr8/core.py:74]: Why do you need the ellipsis object here?

**Review by lhelleckes — COMMENTED** [review]
Good step towards generalisation! Thx

File: calibr8/core.py (hunk 1 above)
Comment (michaelosthege) [reply calibr8/core.py:74]: The array can also be 2-dimensional. For example with a trivariate calibration model:
```
array([[0.11, 0.62],    # concentration
       [0.04, 0.25],    # pH
       [0.58, 0.94]])   # temperature
```
The `arr[..., 0]` with the above array then selects the 0ths entry of the last dimension, so `[0.11, 0.04, 0.58]`.

The same slicing applied to a `(2,)` shaped array such as `[0.1, 0.2]` also gives the 0ths entry of the last dimension (0.1).

In contrast, the slicing without the `Ellipsis` would select the first _row_ of the 2-dimensional array.

*tl;dr:* With the ellipsis-based slicing we can keep ETI/HDI arrays as two-column matrices.

Comment (lhelleckes) [reply calibr8/core.py:74]: Ah, I see.` [:, 0]` would do the same in the 2D case, but would fail in 1D?

File: calibr8/core.py [hunk 2]
```diff
@@ -51,52 +51,111 @@ def __init__(
...
+    def hdi_prob(self) -> float:
+        """ Probability mass of the given highest density interval. """
+        return self._hdi_prob
+
+
+class UnivariateInferenceResult(InferenceResult):
+    """ The result of a numeric infer_independent operation with a univariate model. """
+    def __init__(
+        self,
+        median: float,
```
Comment (lhelleckes) [inline calibr8/core.py:126]: Why is the median here and not in the base class already? You expect examples where the median is not useful?

File: calibr8/core.py (hunk 1 above)
Comment (michaelosthege) [reply calibr8/core.py:74]: Correct

File: calibr8/core.py (hunk 2 above)
Comment (michaelosthege) [reply calibr8/core.py:126]: You can only do a marginal median, so there's no good measure for multivariate data. Taking all the marginal medians is not meaningful if the samples are, for example banana-shaped.
A banana-shape will be the inference result for a bivariate inference with a model that's linear in one and logistic in the other dimension, for example.

So that's why I think we should give a median only for the univariate case.

File: calibr8/core.py [hunk 3]
```diff
@@ -333,63 +495,22 @@ def infer_independent(
...
         -------
-        posterior : NumericPosterior
+        posterior : UnivariateInferenceResult
             the result of the numeric posterior calculation
-        """  
+        """
+        if self.ndim != 1:
+            raise NotImplementedError(
+                f"This calibration model seems to be multivariate (ndim={self.ndim}),"
+                " but does not have an override implementation of .infer_independent()."
```
[the remark this answers is not in the document]
Comment (michaelosthege) [reply calibr8/core.py:504]: I think you misread the string. There's a dot in front, not an underscore?

[the remark this answers is not in the document]
Comment (lhelleckes) [reply calibr8/core.py:504]: Oh, mea culpa

File: calibr8/core.py (hunk 2 above)
Comment (lhelleckes) [reply calibr8/core.py:126]: Approved

**Review by lhelleckes — APPROVED** [review]
Alright, thanks for the additional info. Approved.
